File: packages/validator-go/validator/validate/rule_in.go
package validate

// Membership rule in (docs/spec/validation-rules.md, Values).
//
// Members come from a list (each element as is), a comma-separated string (split
// at U+002C, each item trimmed) or a map (its keys). A value matches a member when
// their canonical texts are the same code points, or when both are decimal numbers
// (numbers, or strings in the decimal grammar) with equal values as doubles.

import (
	"regexp"
	"strconv"
	"strings"

	"github.com/polyspec/crudui/packages/validator-go/validator/compose"
)

// decimalGrammar is the numeric spelling a string member or value may use to
// match by value.
var decimalGrammar = regexp.MustCompile(`^[-+]?([0-9]+\.?[0-9]*|[0-9]*\.?[0-9]+)$`)

// member is one membership value: its canonical text and, when it is a decimal
// number, its value as a double.
type member struct {
	text    string
	number  float64
	numeric bool
}

// newMember builds the member of a scalar value; false when the value is not a
// string, a finite number or a boolean.
func newMember(value any) (member, bool) {
	text, ok := canonicalText(value)
	if !ok {
		return member{}, false
	}
	m := member{text: text}
	switch v := value.(type) {
	case string:
		if decimalGrammar.MatchString(v) {
			m.number, m.numeric = parseDecimal(v), true
		}
	case bool:
	default:
		m.number, m.numeric = numberValue(v)
	}
	return m, true
}

// parseDecimal reads a string in the decimal grammar as the nearest double. A
// spelling beyond the double range reads as an infinity, as ECMAScript reads it.
func parseDecimal(s string) float64 {
	n, _ := strconv.ParseFloat(s, 64)
	return n
}

// inMembers returns the members of an in parameter, or its parameter error.
func inMembers(param any) ([]member, *parameterError) {
	var raw []any
	switch p := param.(type) {
	case []any:
		raw = p
	case string:
		for _, item := range strings.Split(p, ",") {
			raw = append(raw, trimText(item))
		}
	case *compose.OMap:
		for _, key := range p.Keys() {
			raw = append(raw, key)
		}
	case map[string]any:
		for _, key := range sortedKeys(p) {
			raw = append(raw, key)
		}
	default:
		return nil, ruleParameterError("Invalid in parameter: expected a list, a comma-separated string or a map")
	}
	if len(raw) == 0 {
		return nil, ruleParameterError("Invalid in parameter: members must not be empty")
	}
	members := make([]member, 0, len(raw))
	for _, value := range raw {
		m, ok := newMember(value)
		if !ok {
			return nil, ruleParameterError("Invalid in parameter: members must be strings, numbers or booleans")
		}
		if trimText(m.text) == "" {
			return nil, ruleParameterError("Invalid in parameter: members must not be empty")
		}
		members = append(members, m)
	}
	return members, nil
}

// checkIn returns the parameter error of an in parameter.
func checkIn(param any) *parameterError {
	_, err := inMembers(param)
	return err
}

// inMatches reports whether a value is a member. A string value is trimmed; an
// array value matches when every element matches, where an empty element
// (including an empty array or object) passes as an empty value does and any
// other array or object element fails, having no canonical text.
func inMatches(value any, members []member) bool {
	if elements, ok := value.([]any); ok {
		for _, element := range elements {
			if !isEmpty(element) && !scalarMatches(element, members) {
				return false
			}
		}
		return true
	}
	return scalarMatches(value, members)
}

// scalarMatches reports whether a scalar value is a member; a value without
// canonical text is not.
func scalarMatches(value any, members []member) bool {
	if s, ok := value.(string); ok {
		value = trimText(s)
	}
	candidate, ok := newMember(value)
	if !ok {
		return false
	}
	for _, m := range members {
		if candidate.text == m.text || (candidate.numeric && m.numeric && candidate.number == m.number) {
			return true
		}
	}
	return false
}

// ruleIn fails a value that is not a member. Empty values pass.
func ruleIn(value any, ruleParam any, ctx ruleContext) (string, bool) {
	if isEmpty(value) {
		return "", false
	}
	members, _ := inMembers(ruleParam)
	if inMatches(value, members) {
		return "", false
	}
	return ctx.msg("in", "Please select a valid option."), true
}

File: packages/php-ext/src/rule_in.c
#include "engine_internal.h"

#include <stdlib.h>
#include <string.h>

/*
 * Membership (docs/spec/validation-rules.md, "Values"): members come from a list (each element as
 * is), a comma-separated string (each item trimmed) or a map (its keys). A value matches a member
 * when their canonical texts are the same code points, or when both are numeric (numbers, or
 * strings that are numeric text) with equal values.
 */

/* A value or member as the comparison sees it. */
typedef struct {
    ps_text text;
    bool numeric;
    double number;
} member_text;

/* Describe a text as a member: numeric when it is numeric text with a finite value. */
static bool string_member(ps_text text, member_text *member)
{
    *member = (member_text){text, false, 0};
    member->numeric = ps_numeric_text(text, &member->number);
    return true;
}

static bool same_member(const member_text *left, const member_text *right)
{
    return ps_text_equal(left->text, right->text) ||
        (left->numeric && right->numeric && left->number == right->number);
}

static bool scalar_member(const ps_value *value)
{
    return value && (value->kind == PS_STRING || value->kind == PS_INT ||
                     value->kind == PS_FLOAT || value->kind == PS_BOOL);
}

/*
 * Visit every member of a valid parameter in order. The visitor returns 1 to stop with a match,
 * 0 to continue and -1 to stop with a failure. The canonical text of a number or boolean member
 * lives only for the call.
 */
typedef int (*member_visitor)(void *context, const member_text *member);

static int visit_members(const ps_value *parameter, void *context, member_visitor visit)
{
    if (parameter->kind == PS_STRING) {
        ps_text list = ps_string(parameter);
        size_t start = 0;
        for (;;) {
            size_t comma = ps_text_find_byte(list, ',', start);
            size_t end = comma == SIZE_MAX ? list.length : comma;
            member_text member;
            if (!string_member(ps_trim(ps_text_slice(list, start, end)), &member)) return -1;
            int result = visit(context, &member);
            if (result) return result;
            if (comma == SIZE_MAX) return 0;
            start = comma + 1;
        }
    }
    for (size_t i = 0; i < ps_size(parameter); ++i) {
        member_text member;
        if (parameter->kind == PS_OBJECT) {
            if (!string_member(ps_key(parameter, i), &member)) return -1;
            int result = visit(context, &member);
            if (result) return result;
            continue;
        }
        const ps_value *item = ps_at(parameter, i);
        if (item->kind == PS_STRING) {
            if (!string_member(ps_string(item), &member)) return -1;
            int result = visit(context, &member);
            if (result) return result;
            continue;
        }
        ps_chars text;
        if (ps_canonical_text(item, &text) <= 0) return -1;
        member = (member_text){ps_view(text), false, 0};
        if (item->kind != PS_BOOL) member.numeric = ps_numeric_value(item, &member.number);
        int result = visit(context, &member);
        free(text.bytes);
        if (result) return result;
    }
    return 0;
}

static int nonempty_member(void *context, const member_text *member)
{
    (void)context;
    return ps_trim(member->text).length ? 0 : 2;
}

ps_parameter_problem ps_in_parameter(const ps_value *parameter)
{
    static const ps_parameter_problem shape = {.code = "INVALID_RULE_PARAMETER", .message =
        "Invalid in parameter: expected a list, a comma-separated string or a map"};
    static const ps_parameter_problem types = {.code = "INVALID_RULE_PARAMETER", .message =
        "Invalid in parameter: members must be strings, numbers or booleans"};
    static const ps_parameter_problem empty = {.code = "INVALID_RULE_PARAMETER", .message =
        "Invalid in parameter: members must not be empty"};
    static const ps_parameter_problem failed = {.code = "INTERNAL_ERROR", .message = "Validation failed"};
    if (!parameter || (parameter->kind != PS_ARRAY && parameter->kind != PS_OBJECT && parameter->kind != PS_STRING))
        return shape;
    if (parameter->kind != PS_STRING && !ps_size(parameter)) return empty;
    /* Each list element is checked in order: its type, then its text. */
    if (parameter->kind == PS_ARRAY) {
        for (size_t i = 0; i < ps_size(parameter); ++i) {
            const ps_value *item = ps_at(parameter, i);
            if (!scalar_member(item)) return types;
            if (item->kind != PS_STRING) continue;
            if (!ps_trim(ps_string(item)).length) return empty;
        }
        return (ps_parameter_problem){0};
    }
    int result = visit_members(parameter, NULL, nonempty_member);
    if (result < 0) return failed;
    return result ? empty : (ps_parameter_problem){0};
}

typedef struct {
    member_text value;
} match_context;

static int matching_member(void *context, const member_text *member)
{
    return same_member(&((match_context *)context)->value, member) ? 1 : 0;
}

int ps_in_passes(const ps_value *value, const ps_value *parameter)
{
    if (ps_empty_value(value)) return 1;
    /* Every element passes: an empty element as an empty value does, an array or object never. */
    if (value->kind == PS_ARRAY) {
        for (size_t i = 0; i < ps_size(value); ++i) {
            const ps_value *element = ps_at(value, i);
            if (ps_empty_value(element)) continue;
            if (!scalar_member(element)) return 0;
            int result = ps_in_passes(element, parameter);
            if (result <= 0) return result;
        }
        return 1;
    }
    if (!scalar_member(value)) return 0;
    match_context context;
    ps_chars owned = {NULL, 0};
    if (value->kind == PS_STRING) {
        if (!string_member(ps_trim(ps_string(value)), &context.value)) return -1;
    } else {
        if (ps_canonical_text(value, &owned) <= 0) return -1;
        context.value = (member_text){ps_view(owned), false, 0};
        if (value->kind != PS_BOOL) context.value.numeric = ps_numeric_value(value, &context.value.number);
    }
    int result = visit_members(parameter, &context, matching_member);
    free(owned.bytes);
    return result;
}
